source/opt/inst_debug_printf_pass.cpp

```diff
@@ -0,0 +1,257 @@
+    }
+    if (f_ty->width() != 32) {
+      assert(false && "unsupported float width");
+      return;
+    }
+    // Bitcase float32 to uint32
+    Instruction* bc_inst =
+        builder->AddUnaryOp(GetUintId(), SpvOpBitcast, val_inst->result_id());
+    val_ids->push_back(bc_inst->result_id());
+    return;
+  }
+  analysis::Integer* i_ty = val_ty->AsInteger();
+  if (i_ty == nullptr) {
+    assert(false && "unsupported type");
+    return;
+  }
+  if (i_ty->width() == 64) {
+    Instruction* ui64_inst = val_inst;
+    if (i_ty->IsSigned()) {
+      // Bitcast sint64 to uint64
+      ui64_inst = builder->AddUnaryOp(GetUint64Id(), SpvOpBitcast,
+                                      val_inst->result_id());
+    }
+    // Break uint64 into 2x uint32
+    Instruction* lo_ui64_inst =
+        builder->AddUnaryOp(GetUintId(), SpvOpUConvert, ui64_inst->result_id());
+    Instruction* rshift_ui64_inst = builder->AddBinaryOp(
+        GetUint64Id(), SpvOpShiftRightLogical, ui64_inst->result_id(),
+        builder->GetUintConstantId(32));
+    Instruction* hi_ui64_inst = builder->AddUnaryOp(
+        GetUintId(), SpvOpUConvert, rshift_ui64_inst->result_id());
+    val_ids->push_back(lo_ui64_inst->result_id());
+    val_ids->push_back(hi_ui64_inst->result_id());
+    return;
+  }
+  if (i_ty->width() == 8) {
+    Instruction* ui8_inst = val_inst;
+    if (i_ty->IsSigned()) {
+      // Bitcast sint8 to uint8
+      ui8_inst = builder->AddUnaryOp(GetUint8Id(), SpvOpBitcast,
+                                     val_inst->result_id());
+    }
+    // Convert uint8 to uint32
+    Instruction* ui32_inst =
+        builder->AddUnaryOp(GetUintId(), SpvOpUConvert, ui8_inst->result_id());
+    val_ids->push_back(ui32_inst->result_id());
+    return;
+  }
+  if (i_ty->width() != 32) {
+    // TODO(greg-lunarg): Support non-32-bit int
+    assert(false && "unsupported int width");
+    return;
+  }
+  Instruction* ui32_inst = val_inst;
+  if (i_ty->IsSigned()) {
+    // Bitcast sint32 to uint32
+    ui32_inst =
+        builder->AddUnaryOp(GetUintId(), SpvOpBitcast, val_inst->result_id());
+  }
+  // uint32 needs no further processing
+  val_ids->push_back(ui32_inst->result_id());
+}
+
+void InstDebugPrintfPass::GenOutputCode(
+    Instruction* printf_inst, uint32_t stage_idx,
+    std::vector<std::unique_ptr<BasicBlock>>* new_blocks) {
+  BasicBlock* back_blk_ptr = &*new_blocks->back();
+  InstructionBuilder builder(
+      context(), back_blk_ptr,
+      IRContext::kAnalysisDefUse | IRContext::kAnalysisInstrToBlockMapping);
+  // Gen debug printf record validation-specific values. The format string
+  // will have its id written. Vectors will need to be broken down into
+  // component values. float16 will need to be converted to float32. Pointer
+  // and uint64 will need to be converted to two uint32 values. float32 will
+  // need to be bitcast to uint32. int32 will need to be bitcast to uint32.
+  std::vector<uint32_t> val_ids;
+  bool set_seen = false;
```

Comment:
A better name would be `is_first_operand`.

Reply:
k